### ISSUE TITLE
Client blocking connect and close methods return too soon

### ISSUE THREAD

**Issue by dhumeniuk** [TooTallNate/Java-WebSocket#302]
In WebSocketClient, the connect/close latches will count down before the callback is called:

```
    public final void onWebsocketOpen( WebSocket conn, Handshakedata handshake ) {
    connectLatch.countDown();
    onOpen( (ServerHandshake) handshake );
}
```

This means the blocking methods will return before the callback is called. If the consumer of the client interface does something on open that is required before proceeding, there is a race condition where the callback action may not happen soon enough. For example:

```
public class ClientConsumer {
  @Override
  public void onOpen(final ServerHandshake handshake) {
    m_Log.info("Connected to MessageServer: %s", getURI());
    m_Connected = true;
  }

  public connect() {    
    client.connectBlocking();
  }

  public send() {
    if (m_Connected) {
      client.send(...);
    }
  }

ClientConsumer c;
c.connect();  // will return before m_Connect is set possibly
c.send(); // will not always send since m_Connect may still be false
```

This is a simplified example. Our actual wrapper code has more to it.

**Comment by petertorelli** [comment]
This can also be observed in the example client by calling c.send() immediately after c.connect(); calling .isOpen right after .connect() always returns false when testing from the console, but a slight delay solves the problem.

**Comment by gesellix** [comment]
I can confirm the issue. I added some CountDownLatches [in my code](https://github.com/gesellix/docker-client/blob/master/src/integrationTest/groovy/de/gesellix/docker/client/DockerClientImplIntegrationSpec.groovy#L875) to avoid timing issues.

**Comment by rathnaum** [comment]
I am facing the same issue. I guess connectLatch is set to 1 in the Client and probably increasing it to 2 or 3 might solve the issue.

**Comment by marci4** [comment]
Test code:
```java
public class Test {

	public static void main(String[] args) throws URISyntaxException, InterruptedException {
		WebSocketImpl.DEBUG = false;
		while (true) {
			TestClient client = new TestClient();
			client.connectBlocking();
			client.send();
			client.closeBlocking();
			client.send();
		}
	}

	static class TestClient extends WebSocketClient {

		public TestClient() throws URISyntaxException {
			super( new URI( "ws://localhost:8887" ), new Draft_17() );
		}
		public boolean connect = false;

		public boolean closed = false;

		@Override
		public void onOpen( ServerHandshake handshakedata ) {
			int i = Integer.MIN_VALUE;
			while(i != Integer.MAX_VALUE) {
				i++;
			}
			connect = true;
		}

		@Override
		public void onMessage( String message ) {

		}

		@Override
		public void onClose( int code, String reason, boolean remote ) {
			int i = Integer.MIN_VALUE;
			while(i != Integer.MAX_VALUE) {
				i++;
			}
			closed = true;
		}

		@Override
		public void onError( Exception ex ) {

		}

		public void send() {
			if (connect  && !closed) {
				send( "success" );
			} else if (closed) {
				System.out.println("closed");
			} else {
				throw new RuntimeException(  );
			}
		}
	}
}
```

**Comment by marci4** [comment]
This issue should be fixed with the pull request #455 

Greetings
marci4
